Document two JS errors with classes (#32857)
mdn/content: files/en-us/web/javascript/reference/errors/arguments_not_allowed/index.md

---
title: "SyntaxError: arguments is not valid in fields"
slug: Web/JavaScript/Reference/Errors/Arguments_not_allowed
page-type: javascript-error
---

{{jsSidebar("Errors")}}

The JavaScript exception "SyntaxError: arguments is not valid in fields" occurs when the [`arguments`](/en-US/docs/Web/JavaScript/Reference/Functions/arguments) identifier is read in a class field initializer or in a static initialization block, outside of a non-[arrow function](/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions).

## Message

```plain
SyntaxError: 'arguments' is not allowed in class field initializer or static initialization block (V8-based)
SyntaxError: arguments is not valid in fields (Firefox)
SyntaxError: Unexpected identifier 'arguments'. Cannot reference 'arguments' in class field initializer. (Safari)
```

## Error type

{{jsxref("SyntaxError")}}

## What went wrong?

> The `arguments` identifier cannot be read outside of a non-arrow function in a class body or a class static initialization block. Doing so is a syntax error.

Reviewer: ```suggestion
The `arguments` identifier cannot be used in a class field initializer expression or class static initialization block, unless the identifier is immediately contained in a non-arrow function that provides its own `arguments` binding.
```

...just a try. This is hard.

Reviewer 2: I understannndd, and same. There I made another version, I put the explanation short and then added a "Notes" subsection with clarifications. Please let me know your thoughts, and thank you for your time!

---
title: "SyntaxError: arguments is not valid in fields"
slug: Web/JavaScript/Reference/Errors/Arguments_not_allowed
page-type: javascript-error
---

{{jsSidebar("Errors")}}

The JavaScript exception "SyntaxError: arguments is not valid in fields" occurs when the [`arguments`](/en-US/docs/Web/JavaScript/Reference/Functions/arguments) identifier is read in a class field initializer or in a static initialization block, outside of a non-[arrow function](/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions).

## Message

```plain
SyntaxError: 'arguments' is not allowed in class field initializer or static initialization block (V8-based)
SyntaxError: arguments is not valid in fields (Firefox)
SyntaxError: Unexpected identifier 'arguments'. Cannot reference 'arguments' in class field initializer. (Safari)
```

## Error type

{{jsxref("SyntaxError")}}

## What went wrong?

> A class field initializer expression or a class static initialization block cannot read from its scope the `arguments` identifier. Doing so is a syntax error.

Reviewer: ```suggestion
A class field initializer expression or a class static initialization block cannot read the `arguments` of the function that contains the class. Doing so is a syntax error.
```

### Notes

- `arguments` can still be assigned in the above conditions:

  ```js example
  class C {
    arguments = 42;
  }

>   ```

Reviewer: I don't think this is particularly illuminating. You are not "assigning to `arguments`". You are creating a property called `arguments` on class instances. No identifier is restricted from becoming a class property; not even keywords.

Reviewer 2: Very fair, thanks! I took out that note

- This is true even if `arguments` is bound in a parent scope.

> - A non-arrow function declared in the previous conditions will still bind and be able to read its own `arguments` identifier.

Reviewer: @nataliaschlebinger Maybe you know this already, but if it helps you put into context why this error is needed:

Class static blocks and field initializers are actually wrapped into functions. For example:

```js
class A {
  static {
    // ...
  }
}

// Becomes:
class A {
}

(function () {
  // ...
}.call(A));
```

```js
class A {
  foo = this.y;
}

// Becomes:
class A {
  constructor() {
    this.foo = function () {
      return this.y;
    }.call(this);
  }
}
```

This is why the language has to artificially impose this `arguments` restriction here, because you would actually not be able to read `arguments` of the parent, and doing so would be a leak of abstraction.

Maybe it's useful to be put in the description? I'm not sure. But, I think the scoping effects here are still confusing. The only thing to be aware of is that initializers/static blocks basically create their own `arguments` scope boundary, except that in this scope `arguments` is inaccessible.

Reviewer 2: I didn't know this thank you and thanks for the detailed explanation! Still, I don't understand why the explanation for the scoping effects is confusing. Is it talking about "the scope **of** an expression"? By that I meant the scope that the expression can "see".

I don't wanna talk about mixing classes and functions, because I don't want people to think of the times they've heard that "classes are just syntactic sugar for constructor functions", I would like them to think of classes as a thing in their own right. (Maybe I'm wrong!!)

Do you think it's necessary to talk about this? I'm questioning because it's a thing the language does internally, and only the restriction is exposed to the programmers. If it did this internally but with no restriction, `arguments` in our case would just always be empty, for example. If it is necessary, I'll write it in no problem!

Thank you for your time.

## Examples

```js example-bad
function makeOne() {
    class C {
        args = { ...arguments }; // SyntaxError: arguments is not valid in fields
    }
    return new C();

> }

Reviewer: While it's invalid code, you still need to make sure it's well-formatted. For us, it means using 2-space indentation.

Reviewer 2: Ah, thank you!! There, I fixed it!

```

```js example-bad
let CArgs;
class C {
    static {
        CArgs = arguments; // SyntaxError: arguments is not valid in fields
    }
}
```

```js example-good
class C {
  args = {};
  constructor() {

>     this.args = arguments; // You can use arguments in methods

Reviewer: This is a constructor, not a method.

Reviewer 2: Right now we have documented on MDN that the constructor is `a special method`, and the spec seems to treat it like that? Do you think that in any case it's clearer to use a regular method for the example?

Reviewer: Constructors have different syntax restrictions from methods in many regards (such as no `async`, can use `super()`...) So it's better to not conflate them when talking about syntax.

Reviewer 2: I understand, I think I fixed it there!
